libidn-1.15/doc/man/stringprep.3
.\" DO NOT MODIFY THIS FILE!  It was generated by gdoc.
.TH "stringprep" 3 "1.15" "libidn" "libidn"
.SH NAME
stringprep \- prepare internationalized string
.SH SYNOPSIS
.B #include <stringprep.h>
.sp
.BI "int stringprep(char * " in ", size_t " maxlen ", Stringprep_profile_flags " flags ", const Stringprep_profile * " profile ");"
.SH ARGUMENTS
.IP "char * in" 12
input/ouput array with string to prepare.
.IP "size_t maxlen" 12
maximum length of input/output array.
.IP "Stringprep_profile_flags flags" 12
a \fBStringprep_profile_flags\fP value, or 0.
.IP "const Stringprep_profile * profile" 12
pointer to \fBStringprep_profile\fP to use.
.SH "DESCRIPTION"
Prepare the input zero terminated UTF\-8 string according to the
stringprep profile, and write back the result to the input string.

Note that you must convert strings entered in the systems locale
into UTF\-8 before using this function, see
\fBstringprep_locale_to_utf8()\fP.

Since the stringprep operation can expand the string, \fImaxlen\fP
indicate how large the buffer holding the string is.  This function
will not read or write to characters outside that size.

The \fIflags\fP are one of \fBStringprep_profile_flags\fP values, or 0.

The \fIprofile\fP contain the \fBStringprep_profile\fP instructions to
perform.  Your application can define new profiles, possibly
re\-using the generic stringprep tables that always will be part of
the library, or use one of the currently supported profiles.
.SH "RETURN VALUE"
Returns \fBSTRINGPREP_OK\fP iff successful, or an error code.
.SH "REPORTING BUGS"
Report bugs to <bug-libidn@gnu.org>.
GNU Libidn home page: http://www.gnu.org/software/libidn/
General help using GNU software: http://www.gnu.org/gethelp/
.SH COPYRIGHT
Copyright \(co 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009 Simon Josefsson.
.br
Copying and distribution of this file, with or without modification,
are permitted in any medium without royalty provided the copyright
notice and this notice are preserved.
.SH "SEE ALSO"
The full documentation for
.B libidn
is maintained as a Texinfo manual.  If the
.B info
and
.B libidn
programs are properly installed at your site, the command
.IP
.B info libidn
.PP
should give you access to the complete manual.


noflushd/NEWS
* noflushd 2.7.5 (09. May 2005)

  - Fix handling of raid and similar meta devices on 2.6 kernels.

* noflushd 2.7.4 (08. Aug 2004)

  - Fix lousy spindown times on 2.6 kernels.
  - Fix SCSI support on 2.6 kernels.
  - Fix pid detection in generic startup script.
  - Document quirks with hotplugged drives.

* noflushd 2.7.3 (09. Apr 2004)

  - Fix race condition that might leave kupdated/pdflushd disabled
    aftern noflushd has terminated.
  - Fix spindown when all disks are auto-detected.

* noflushd 2.7.2 (16. Mar 2004)

  - A change in 2.7.1 erroneously flushed the buffer cache before
    spindown, causing frequent spinups. This change has been reverted.
    
* noflushd 2.7.1 (07. Mar 2004)

  - Fix immediate spinup on 2.6 kernels.
  - Elevate SCSI support from compile-time to command line option.
  - Bug fixes in the disk statistics parser for 2.6 kernels.
  - Ship kernel patch to track down which processes cause spinups.

* noflushd 2.7 (20. Jan 2004)

  - Support Linux kernel 2.6.x.
  - Default to monitorring all disks unless specific disks are given
    on the command line. This change should yield a sane default
    behaviour for a majority of users without requiring explicit
    configuration.

* noflushd 2.6.3 (13. May 2002)

  - Bugfix for arm and other archs with weird /proc/partitions layout.

* noflushd 2.6.2 (20. Apr 2002)

  - Add IDE raid support.
  - Add timeout logging script.

* noflushd 2.6.1 (03. Feb. 2002)

  - Fix BUG when using noflushd with disks on second IDE channel on 2.4
    kernels.

* noflushd 2.6 (22. Jan 2002)

  - Document noflushd's suckage with any journaling fs, including ext3.
  - New feature prevents spindown if machine is in interactive use.
  - Add hack fix some '0 spindown time' uglies.

* noflushd 2.5 (06. Aug 2001)

  - Make sure hotplugged disks get synced properly.
  - Only sync present partitions. (Fixes high CPU usage bug on devfs
    systems.)
  - Generic startup skript works on latest 2.4 kernels.

* noflushd 2.4 (13. Mar 2001)

  - Timeout lists are back to a working state.
  - Support very latest 2.4 kernels.

* noflushd 2.3 (12. Jan 2001)

  - Support latest 2.4 kernels. (Leaves kernels 2.3.99pre8 to 2.4.0-test8
    unsupported.)
  - Fix multi-disk support with long bdflush timeouts.

* noflushd 2.2 (22. Dec 2000)

  - Fixed bug in spindown code. Now actually spins down. Duh!

* noflushd 2.1 (20. Dec 2000)

  - Disks get synced prior to spindown.
  - Improved devfs support.

* noflushd 2.0 (14. Nov 2000)

  - noflushd now comes with its own syncer that takes over the task from
    kupdate when the first read-write mounted disk is spun down. This
    enhances spindown times in a multidisk environment at the expense of
    less smooth and more bursty background flushing when the noflushd 
    syncer is in use.

  - Conflicting entries in 2.2-style /proc/stat (e.g. /dev/sda and /dev/hda)
    won't cause noflushd to give up anymore. Instead, noflushd will treat
    both disks pretty much like one, i.e. they're either both up or spun
    down, no matter if activity only goes to one of them.

  - Compilation environment switched to automake/autoconf.

  - When installing via 'make install', noflushd is no longer added to
    any runlevel. 

  - noflushd moved to SourceForge. The project homepage is located at
    http://noflushd.sourceforge.net, a mailing list is available via
    noflushd-devel@lists.sourceforge.net. Visit the homepage for 
    subscription information.



libidn-1.15/csharp/Stringprep.cs
/// <summary>
/// Copyright (C) 2004, 2005, 2006, 2007, 2008, 2009  Free Software Foundation, Inc.
/// *
/// Author: Alexander Gnauck AG-Software, mailto:gnauck@ag-software.de
/// *
/// This file is part of GNU Libidn.
/// *
/// This library is free software; you can redistribute it and/or
/// modify it under the terms of the GNU Lesser General Public License
/// as published by the Free Software Foundation; either version 2.1 of
/// the License, or (at your option) any later version.
/// *
/// This library is distributed in the hope that it will be useful, but
/// WITHOUT ANY WARRANTY; without even the implied warranty of
/// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
/// Lesser General Public License for more details.
/// *
/// You should have received a copy of the GNU Lesser General Public
/// License along with this library; if not, write to the Free Software
/// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301
/// USA
/// </summary>

using System;
using System.Text;

namespace Gnu.Inet.Encoding
{	
	/// <summary> This class offers static methods for preparing internationalized
	/// strings. It supports the following stringprep profiles:
	/// <ul>
	/// <li>RFC3491 nameprep
	/// <li>RFC3920 XMPP nodeprep and resourceprep
	/// </ul>
	/// Note that this implementation only supports 16-bit Unicode code
	/// points.
	/// </summary>
	public class Stringprep
	{
		/// <summary> Preps a name according to the Stringprep profile defined in
		/// RFC3491. Unassigned code points are not allowed.
		/// *
		/// </summary>
		/// <param name="input">the name to prep.
		/// </param>
		/// <returns> the prepped name.
		/// @throws StringprepException If the name cannot be prepped with
		/// this profile.
		/// @throws NullPointerException If the name is null.
		/// 
		/// </returns>
		public static string NamePrep(string input)
		{
			return NamePrep(input, false);
		}
		
		/// <summary> Preps a name according to the Stringprep profile defined in
		/// RFC3491.
		/// *
		/// </summary>
		/// <param name="input">the name to prep.
		/// </param>
		/// <param name="allowUnassigned">true if the name may contain unassigned
		/// code points.
		/// </param>
		/// <returns> the prepped name.
		/// @throws StringprepException If the name cannot be prepped with
		/// this profile.
		/// @throws NullPointerException If the name is null.
		/// 
		/// </returns>
		public static string NamePrep(string input, bool allowUnassigned)
		{
			if (input == null)
			{
				throw new System.NullReferenceException();
			}
			
			StringBuilder s = new StringBuilder(input);
			
			if (!allowUnassigned && Contains(s, RFC3454.A1))
			{
				throw new StringprepException(StringprepException.CONTAINS_UNASSIGNED);
			}
			
			Filter(s, RFC3454.B1);
			Map(s, RFC3454.B2search, RFC3454.B2replace);
			
			s = new StringBuilder(NFKC.NormalizeNFKC(s.ToString()));
			// B.3 is only needed if NFKC is not used, right?
			// map(s, RFC3454.B3search, RFC3454.B3replace);
			
			if (Contains(s, RFC3454.C12) || Contains(s, RFC3454.C22) || Contains(s, RFC3454.C3) || Contains(s, RFC3454.C4) || Contains(s, RFC3454.C5) || Contains(s, RFC3454.C6) || Contains(s, RFC3454.C7) || Contains(s, RFC3454.C8))
			{
				// Table C.9 only contains code points > 0xFFFF which Java
				// doesn't handle
				throw new StringprepException(StringprepException.CONTAINS_PROHIBITED);
			}
			
			// Bidi handling
			bool r = Contains(s, RFC3454.D1);
			bool l = Contains(s, RFC3454.D2);
			
			// RFC 3454, section 6, requirement 1: already handled above (table C.8)
			
			// RFC 3454, section 6, requirement 2
			if (r && l)
			{
				throw new StringprepException(StringprepException.BIDI_BOTHRAL);
			}
			
			// RFC 3454, section 6, requirement 3
			if (r)
			{
				if (!Contains(s[0], RFC3454.D1) || !Contains(s[s.Length - 1], RFC3454.D1))
				{
					throw new StringprepException(StringprepException.BIDI_LTRAL);
				}
			}
			
			return s.ToString();
		}
		
		/**
		* Characters prohibited by RFC3920 nodeprep that aren't defined as
		* part of the RFC3454 tables.
		*/
		private static char [] RFC3920_NODEPREP_PROHIBIT = new char [] {
				 '\u0022', '\u0026', '\'',     '\u002F',
				 '\u003A', '\u003C', '\u003E', '\u0040'
		 };
		
		/// <summary> Preps a node name according to the Stringprep profile defined in
		/// RFC3920. Unassigned code points are not allowed.
		/// *
		/// </summary>
		/// <param name="input">the node name to prep.
		/// </param>
		/// <returns> the prepped node name.
		/// @throws StringprepException If the node name cannot be prepped
		/// with this profile.
		/// @throws NullPointerException If the node name is null.
		/// 
		/// </returns>
		public static string NodePrep(string input)
		{
			return NodePrep(input, false);
		}
		
		/// <summary>
        /// Preps a node name according to the Stringprep profile defined in RFC3920.
		/// </summary>
		/// <param name="input">the node name to prep.
		/// </param>
		/// <param name="allowUnassigned">true if the node name may contain
		/// unassigned code points.
		/// </param>
		/// <returns> the prepped node name.
		/// @throws StringprepException If the node name cannot be prepped
		/// with this profile.
		/// @throws NullPointerException If the node name is null.
		/// 
		/// </returns>
		public static string NodePrep(string input, bool allowUnassigned)
		{
			if (input == null)
			{
				throw new System.NullReferenceException();
			}
			
			StringBuilder s = new StringBuilder(input);
			
			if (!allowUnassigned && Contains(s, RFC3454.A1))
			{
				throw new StringprepException(StringprepException.CONTAINS_UNASSIGNED);
			}
			
			Filter(s, RFC3454.B1);
			Map(s, RFC3454.B2search, RFC3454.B2replace);
			
			s = new StringBuilder(NFKC.NormalizeNFKC(s.ToString()));
			
			if (Contains(s, RFC3454.C11) || Contains(s, RFC3454.C12) || Contains(s, RFC3454.C21) || Contains(s, RFC3454.C22) || Contains(s, RFC3454.C3) || Contains(s, RFC3454.C4) || Contains(s, RFC3454.C5) || Contains(s, RFC3454.C6) || Contains(s, RFC3454.C7) || Contains(s, RFC3454.C8) || Contains(s, RFC3920_NODEPREP_PROHIBIT))
			{
				// Table C.9 only contains code points > 0xFFFF which Java
				// doesn't handle
				throw new StringprepException(StringprepException.CONTAINS_PROHIBITED);
			}
			
			// Bidi handling
			bool r = Contains(s, RFC3454.D1);
			bool l = Contains(s, RFC3454.D2);
			
			// RFC 3454, section 6, requirement 1: already handled above (table C.8)
			
			// RFC 3454, section 6, requirement 2
			if (r && l)
			{
				throw new StringprepException(StringprepException.BIDI_BOTHRAL);
			}
			
			// RFC 3454, section 6, requirement 3
			if (r)
			{
				if (!Contains(s[0], RFC3454.D1) || !Contains(s[s.Length - 1], RFC3454.D1))
				{
					throw new StringprepException(StringprepException.BIDI_LTRAL);
				}
			}
			
			return s.ToString();
		}
		
		/// <summary>
        /// Preps a resource name according to the Stringprep profile defined
		/// in RFC3920. Unassigned code points are not allowed.
		/// </summary>
		/// <param name="input">the resource name to prep.
		/// </param>
		/// <returns> the prepped node name.
		/// @throws StringprepException If the resource name cannot be prepped
		/// with this profile.
		/// @throws NullPointerException If the resource name is null.
		/// 
		/// </returns>
		public static string ResourcePrep(string input)
		{
			return ResourcePrep(input, false);
		}
		
		/// <summary>
        /// Preps a resource name according to the Stringprep profile defined
		/// in RFC3920.
		/// </summary>
		/// <param name="input">the resource name to prep.
		/// </param>
		/// <param name="allowUnassigned">true if the resource name may contain
		/// unassigned code points.
		/// </param>
		/// <returns>
        /// the prepped node name.
		/// @throws StringprepException If the resource name cannot be prepped
		/// with this profile.
		/// @throws NullPointerException If the resource name is null.
		/// 
		/// </returns>        
		public static string ResourcePrep(string input, bool allowUnassigned)
		{
			if (input == null)
			{
				throw new System.NullReferenceException();
			}
			
			StringBuilder s = new StringBuilder(input);
			
			if (!allowUnassigned && Contains(s, RFC3454.A1))
			{
				throw new StringprepException(StringprepException.CONTAINS_UNASSIGNED);
			}
			
			Filter(s, RFC3454.B1);
			
			s = new StringBuilder(NFKC.NormalizeNFKC(s.ToString()));
			
			if (Contains(s, RFC3454.C12) || Contains(s, RFC3454.C21) || Contains(s, RFC3454.C22) || Contains(s, RFC3454.C3) || Contains(s, RFC3454.C4) || Contains(s, RFC3454.C5) || Contains(s, RFC3454.C6) || Contains(s, RFC3454.C7) || Contains(s, RFC3454.C8))
			{
				// Table C.9 only contains code points > 0xFFFF which Java
				// doesn't handle
				throw new StringprepException(StringprepException.CONTAINS_PROHIBITED);
			}
			
			// Bidi handling
			bool r = Contains(s, RFC3454.D1);
			bool l = Contains(s, RFC3454.D2);
			
			// RFC 3454, section 6, requirement 1: already handled above (table C.8)
			
			// RFC 3454, section 6, requirement 2
			if (r && l)
			{
				throw new StringprepException(StringprepException.BIDI_BOTHRAL);
			}
			
			// RFC 3454, section 6, requirement 3
			if (r)
			{
				if (!Contains(s[0], RFC3454.D1) || !Contains(s[s.Length - 1], RFC3454.D1))
				{
					throw new StringprepException(StringprepException.BIDI_LTRAL);
				}
			}
			
			return s.ToString();
		}
		
		internal static bool Contains(StringBuilder s, char[] p)
		{
			for (int i = 0; i < p.Length; i++)
			{
				char c = p[i];
				for (int j = 0; j < s.Length; j++)
				{
					if (c == s[j])
					{
						return true;
					}
				}
			}
			return false;
		}
		
		internal static bool Contains(StringBuilder s, char[][] p)
		{
			for (int i = 0; i < p.Length; i++)
			{
				char[] r = p[i];
				if (1 == r.Length)
				{
					char c = r[0];
					for (int j = 0; j < s.Length; j++)
					{
						if (c == s[j])
						{
							return true;
						}
					}
				}
				else if (2 == r.Length)
				{
					char f = r[0];
					char t = r[1];
					for (int j = 0; j < s.Length; j++)
					{
						if (f <= s[j] && t >= s[j])
						{
							return true;
						}
					}
				}
			}
			return false;
		}
		
		internal static bool Contains(char c, char[][] p)
		{
			for (int i = 0; i < p.Length; i++)
			{
				char[] r = p[i];
				if (1 == r.Length)
				{
					if (c == r[0])
					{
						return true;
					}
				}
				else if (2 == r.Length)
				{
					char f = r[0];
					char t = r[1];
					if (f <= c && t >= c)
					{
						return true;
					}
				}
			}
			return false;
		}
		
		internal static void Filter(StringBuilder s, char[] f)
		{
			for (int i = 0; i < f.Length; i++)
			{
				char c = f[i];
				
				int j = 0;
				while (j < s.Length)
				{
					if (c == s[j])
					{
						//s.deleteCharAt(j);
						s.Remove(j, 1);
					}
					else
					{
						j++;
					}
				}
			}
		}
		
		internal static void Filter(StringBuilder s, char[][] f)
		{
			for (int i = 0; i < f.Length; i++)
			{
				char[] r = f[i];
				
				if (1 == r.Length)
				{
					char c = r[0];
					
					int j = 0;
					while (j < s.Length)
					{
						if (c == s[j])
						{
							//s.deleteCharAt(j);
							s.Remove(j, 1);
						}
						else
						{
							j++;
						}
					}
				}
				else if (2 == r.Length)
				{
					char from = r[0];
					char to = r[1];
					
					int j = 0;
					while (j < s.Length)
					{
						if (from <= s[j] && to >= s[j])
						{
							//s.deleteCharAt(j);
							s.Remove(j, 1);
						}
						else
						{
							j++;
						}
					}
				}
			}
		}
		
		internal static void Map(StringBuilder s, char[] search, string[] replace)
		{
			for (int i = 0; i < search.Length; i++)
			{
				char c = search[i];
				
				int j = 0;
				while (j < s.Length)
				{
					if (c == s[j])
					{
						//s.deleteCharAt(j);
						s.Remove(j, 1);
						if (null != replace[i])
						{
							s.Insert(j, replace[i]);
							j += replace[i].Length - 1;
						}
					}
					else
					{
						j++;
					}
				}
			}
		}
	}
}